### PR TITLE
Flexible custom authentication exceptions configuration

### LINKED ISSUE #1681: Add ability to easily add custom authentication exceptions for custom error messages
Currently, the `AuthenticationExceptionHandler` is not designed for easily adding custom authentication exceptions and thus making them available for custom UI message mapping in `messages.properties`.

Refactor `AuthenticationExceptionHandler` to add such facility for adding custom authentication exceptions to the list of the default ones and make it work via `cas.properties`, so the deployer would only need to do 2 simple things (well, besides introducing custom `GeneralSecurityException` or `PreventedException` and throwing them from their custom authn handlers):

1) Add a list of such exceptions to `cas.properties`:

`cas.custom.authentication.exceptions=org.example.MyCustomEx1,org.example.MyCustomEx2`

2) Map the messages for those in `messages.properties`:

```properties
authenticationFailure.MyCustomEx1=Custom msg1
authenticationFailure.MyCustomEx2=Custom msg2
...
```

### PR DESCRIPTION
**Pull request by dima767**
Closes #1681

### CODE REVIEW

File: cas-server-core-authentication/src/test/java/org/jasig/cas/authentication/FileTrustStoreSslSocketFactoryTests.java [hunk 1]
```diff
@@ -18,7 +19,8 @@
 public class FileTrustStoreSslSocketFactoryTests {
 
     @Test
-     public void verifyTrustStoreLoadingSuccessfullyWithCertAvailable() throws Exception {
+    @Ignore
```
Comment (mmoayyed) [inline cas-server-core-authentication/src/test/java/org/jasig/cas/authentication/FileTrustStoreSslSocketFactoryTests.java]: You should turn this back on.

File: cas-server-core-authentication/src/main/java/org/jasig/cas/web/flow/AuthenticationExceptionHandler.java [hunk 2]
```diff
@@ -74,17 +79,46 @@
...
      *
      * @param errors List of errors in order of descending precedence.
      */
+    @Value("#{'${cas.custom.authentication.exceptions:}'.split(',')}")
     public void setErrors(final List<Class<? extends Exception>> errors) {
-        this.errors = errors;
+        //The specifics of the default empty value: this results in the list with one null element. So just get rid of null and have
+        //an empty list as a result.
+        final List<Class<? extends Exception>> nonNullErrors = errors.stream().filter(Objects::nonNull).collect(toList());
+        if(nonNullErrors.size() == 0) {
```
Comment (mmoayyed) [inline cas-server-core-authentication/src/main/java/org/jasig/cas/web/flow/AuthenticationExceptionHandler.java]: Checkstyle will likely error out, that this should be `if (nonNullErrors.isEmpty()) {

**Comment by mmoayyed** [comment]
Looks great. Few minor comments. 

Also, could we get this behavior documented somewhere?

File: cas-server-core-authentication/src/main/java/org/jasig/cas/web/flow/AuthenticationExceptionHandler.java (hunk 2 above)
Comment (dima767) [inline cas-server-core-authentication/src/main/java/org/jasig/cas/web/flow/AuthenticationExceptionHandler.java]: Nope. Checkstyle is happy. Ran the entire CAS server build (with all the tests) locally before pushing. ;-)

File: cas-server-core-authentication/src/test/java/org/jasig/cas/authentication/FileTrustStoreSslSocketFactoryTests.java (hunk 1 above)
Comment (dima767) [inline cas-server-core-authentication/src/test/java/org/jasig/cas/authentication/FileTrustStoreSslSocketFactoryTests.java]: It WILL FAIL

Comment (mmoayyed) [inline cas-server-core-authentication/src/test/java/org/jasig/cas/authentication/FileTrustStoreSslSocketFactoryTests.java]: Not according to Travis. Not sure why it fails locally, but Travis runs it fine, right?

File: cas-server-core-authentication/src/main/java/org/jasig/cas/web/flow/AuthenticationExceptionHandler.java (hunk 2 above)
Comment (mmoayyed) [inline cas-server-core-authentication/src/main/java/org/jasig/cas/web/flow/AuthenticationExceptionHandler.java]: OK, but that is what you want to use instead right?

Comment (mmoayyed) [inline cas-server-core-authentication/src/main/java/org/jasig/cas/web/flow/AuthenticationExceptionHandler.java]: Because you're duplicating isEmpty.

**Comment by dima767** [comment]
wrt documentation - please suggest a place in our docs and I will do

File: cas-server-core-authentication/src/test/java/org/jasig/cas/authentication/FileTrustStoreSslSocketFactoryTests.java (hunk 1 above)
Comment (dima767) [inline cas-server-core-authentication/src/test/java/org/jasig/cas/authentication/FileTrustStoreSslSocketFactoryTests.java]: I don't know. I haven't pushed it with it enabled to find out. Well, anyway, let me enable it, push it and then we'll see

File: cas-server-core-authentication/src/main/java/org/jasig/cas/web/flow/AuthenticationExceptionHandler.java (hunk 2 above)
Comment (dima767) [inline cas-server-core-authentication/src/main/java/org/jasig/cas/web/flow/AuthenticationExceptionHandler.java]: Who cares? ;-) OK, I'll make your day - will change it ;-)

**Comment by mmoayyed** [comment]
I suppose for docs you could choose the UI customizations doc or the webflow config docs. Both seem like good candidates to me. Webflow might be more appropriate for this kind of thing. They are all inside the docs module of course under the installation folder. Thanks!

**Comment by dima767** [comment]
Will add the documentation tomorrow

**Comment by dima767** [comment]
OK, I've added the docs

**Comment by mmoayyed** [comment]
Awesome. Thanks :+1:

**Pull request merged**